internal/mesh/internal/controllers/explicitdestinations/controller.go

```diff
@@ -0,0 +1,300 @@
+	destinationIDs := r.mapper.DestinationsForWorkload(req.ID)
+	rt.Logger.Trace("cached destinations IDs", "ids", destinationIDs)
+
+	decodedDestinations, err := r.fetchDestinations(ctx, rt.Client, destinationIDs)
+	if err != nil {
+		rt.Logger.Error("error fetching mapper", "error", err)
+		return err
+	}
+
+	if len(decodedDestinations) > 0 {
+		r.mapper.TrackDestinations(req.ID, decodedDestinations)
+	} else {
+		r.mapper.UntrackComputedExplicitDestinations(req.ID)
+	}
+
+	duplicates := findConflicts(decodedDestinations)
+
+	newComputedDestinationsData := &pbmesh.ComputedExplicitDestinations{}
+	for _, dst := range decodedDestinations {
+		updatedStatus := &pbresource.Status{
+			ObservedGeneration: dst.GetResource().GetGeneration(),
+		}
+
+		// First check if this resource has a conflict. If it does, update status and move to the next resource.
+		if _, ok := duplicates[resource.NewReferenceKey(dst.GetResource().GetId())]; ok {
+			rt.Logger.Trace("skipping this Destinations resource because it has conflicts with others", "id", dst.GetResource().GetId())
+			updatedStatus.Conditions = append(updatedStatus.Conditions, ConditionConflictFound(workload.GetResource().GetId()))
+		} else {
+			valid, cond := validate(ctx, rt.Client, dst)
+
+			// Only add it to computed mapper if its mapper are valid.
+			if valid {
+				newComputedDestinationsData.Destinations = append(newComputedDestinationsData.Destinations, dst.GetData().GetDestinations()...)
+			} else {
+				rt.Logger.Trace("Destinations is not valid", "condition", cond)
+			}
+
+			updatedStatus.Conditions = append(updatedStatus.Conditions, ConditionConflictNotFound, cond)
+		}
+
+		// Write status for this destination.
+		currentStatus := dst.GetResource().GetStatus()[ControllerName]
+
+		// If the status is unchanged then we should return and avoid the unnecessary write
+		if !resource.EqualStatus(currentStatus, updatedStatus, false) {
+			rt.Logger.Trace("updating status", "id", dst.GetResource().GetId())
+			_, err = rt.Client.WriteStatus(ctx, &pbresource.WriteStatusRequest{
+				Id:     dst.GetResource().GetId(),
+				Key:    ControllerName,
+				Status: updatedStatus,
+			})
+			if err != nil {
+				rt.Logger.Error("error writing new status", "id", dst.GetResource().GetId(), "error", err)
+				return err
+			}
+		}
+	}
+
+	// If after fetching and validating, we don't have any destinations,
+	// we need to skip reconcile and delete the resource.
+	if len(newComputedDestinationsData.GetDestinations()) == 0 {
+		rt.Logger.Trace("found no destinations associated with this workload")
+
+		if ced != nil {
+			rt.Logger.Trace("deleting ComputedDestinations")
+			_, err = rt.Client.Delete(ctx, &pbresource.DeleteRequest{Id: req.ID})
+			if err != nil {
+				// If there's an error deleting CD, we want to re-trigger reconcile again.
+				rt.Logger.Error("error deleting ComputedExplicitDestinations", "error", err)
+				return err
+			}
+		}
+
+		return nil
+	}
+
```

Comment:
Should these destinations be deterministically sorted so the `proto.Equal` below doesn't fail randomly?

Reply:
Yep, added alphabetical sorting